Introduce GDAKI with DOCA GPUNetIO + DOCA RDMA, stream mode

e-ago opened the pull request.

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,391 @@
+      *connection_notif; /* The RDMA_CM connection instance */
+};
+
+class nixlDocaEngine : public nixlBackendEngine {
```
I suggest to move the code before this line to a separate _aux.h to make this file shorter, hard to read through.

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,391 @@
+  ~nixlDocaEngine();
+
+  bool supportsRemote() const { return true; }
+  bool supportsLocal() const { return false; }
```
We can't do between 2 GPUs on the same machine? Maybe add a todo "to be supported in next release" or something like that. I think that's something expected.

mkhazraee replied:
The priority should be NVLINK, if not available use a NIC. (That's UCX flow now). I guess I can see PCIe even as a next level last resort, but less important, unlikely to get to such a system without any NICs.

I think for now we can keep it as TODO, but something important to add, as someone who is using this plugin for sure wants to communicate within the NVLINK domain too.

e-ago replied:
Makes sense, let's keep it a TODO for next releases

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,391 @@
+                            nixlBackendMD *&out) override;
+  nixl_status_t deregisterMem(nixlBackendMD *meta) override;
+
+  nixl_status_t loadLocalMD(nixlBackendMD *input,
```
If you don't support local, then this is redundant.

e-ago replied:
see my previous comment

mkhazraee replied:
If we add the TODO for now, we can remove this.

mkhazraee replied:
I mean as of now that supportsLocal is false, this should be removed. If you add the todo in supportsLocal to say we will add support for it that would be good, but anyways this should be removed.

e-ago replied:
removed for now

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,391 @@
+                                      uint32_t permissions, void *addr,
+                                      uint32_t size, nixlDocaMem *mem);
+
+#if __cplusplus
```
Why do we need this? As if it's compile in C mode? We can't do that for the lib anyways, we use a lot of STL. Or something to do with CUDA?

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,391 @@
+
+  pthread_t server_thread_id;
+
+  class nixlDocaBckndReq : public nixlLinkElem<nixlDocaBckndReq>,
```
nixlLinkElm is something Artem developed specifically for UCX, not necessary to use, unless it made your life easier.

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,391 @@
+  mutable std::atomic<uint32_t> xferRingPos;
+  mutable uint32_t firstXferRingPos;
+
+  struct docaXferCompletion *completion_list_gpu;
```
Maybe we can create an internal class for these two, reducing number of members per class. This is the main gpunetio backend engine and has several members, giving it a tad of organization. For now maybe just a TODO in the code would do for future.

mkhazraee commented on src/plugins/doca/doca_plugin.cpp:
```diff
@@ -0,0 +1,83 @@
+
+// Plugin version information
+static const char *PLUGIN_NAME = "DOCA";
+static const char *PLUGIN_VERSION = "0.1.0";
```
@aranadive What's the current versioning rule?

mkhazraee commented on src/plugins/doca/doca_plugin.cpp:
```diff
@@ -0,0 +1,83 @@
+// Function to get supported backend mem types
+static nixl_mem_list_t get_backend_mems() {
+  nixl_mem_list_t mems;
+  mems.push_back(DRAM_SEG);
```
Just double checking, does it make sense to have gpu-initiated transfers that go to CPU? I was thinking gpu net io would only make sense for GPU to GPU, so only VRAM seg should be here. But if it makes sense, all good.

e-ago replied:
Yes it may make sense. If an application needs to modify and send from a CUDA kernel data allocated in CPU pinned memory, the application can still use GPU communications for it.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+ * Memory management
+ *****************************************/
+nixl_status_t nixlDocaEngine::registerMem(const nixlBlobDesc &mem,
+										  const nixl_mem_t &nixl_mem,
```
Also nixl_mem is not being used, it should be checked if it's VRAM seg to be within GPU memories, if DRAM seg not to be within GPU memory, and if other seg to error (if dram doesn't make sense as my other comment, then only checking for VRAM or error out). I recall for UCX we had to do some figuring out about cuda context too. Do we support multi-GPU agent now in this Plugin?

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+}
+
+nixl_status_t nixlDocaEngine::connect(const std::string &remote_agent) {
+	/* Already connected to remote QP at loadRemoteConnInfo time */
```
Let's add a TODO here and in disconnect too, that part of code from function X would be moved here.

e-ago replied:
Done

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+
+nixl_mem_list_t nixlDocaEngine::getSupportedMems() const {
+	nixl_mem_list_t mems;
+	mems.push_back(DRAM_SEG);
```
Similar comments as the doca_plugin.cpp one, does going to host memory with GPU Net IO make sense in the first place? Or only GPU to GPU makes sense?

e-ago replied:
[Previous answer](https://github.com/ai-dynamo/nixl/pull/404#discussion_r2119213670)

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+	}
+	std::cout << std::endl;
+
+	if (custom_params->count("gpu_devices") == 0)
```
In UCX we would find the GPU device based on the registered memory. Do we really need this?

mkhazraee replied:
I see you only support one, we did the check in register and figured the GPU device. I think that makes things easier, not to require any parameter here.

mkhazraee replied:
I see the nuance, and then in registration if it fails the whole plugin should fail. Or there is a middle ground you can check for existence of GPUs, but not get a specific one?

If there is no other way around and we need the specific GPU during init, let's default to GPU0 if none was provided, many times they run it within k8s with single GPU.

e-ago replied:
or maybe, if null, I can retrieve the cuda id from current context at this point. But this may be an issue if no cuInit has been invoked before agent creation.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+		std::cout << "cuda_id " << cuda_id << "\n";
+	}
+	std::cout << std::endl;
+
```
I assumed there should be an optional parameter for number of cuda streams, if user set it to 0, it means they would provide it per transfer. If not, let's set it to some default value.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+	return NIXL_ERR_BACKEND;
+}
+
+nixl_status_t nixlDocaEngine::loadLocalMD(nixlBackendMD *input,
```
Oh I see, you can remove it, that's why this method was not pure virtual. Agent will never call this method if a backend doesn't support local, and in agent the default implementation is to return error if in future some code was causing such erroneous scenario. Let's not override it with SUCCESS.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+										   const nixl_mem_t &nixl_mem,
+										   const std::string &remote_agent,
+										   nixlBackendMD *&output) {
+	return internalMDHelper(input.metaInfo, remote_agent, output);
```
Just reminder for our talks, to bring the code here as you said and mention the TODO.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+}
+
+nixl_status_t nixlDocaEngine::releaseReqH(nixlBackendReqH *handle) const {
+	firstXferRingPos = xferRingPos.load() & (DOCA_XFER_REQ_MAX - 1);
```
Do we support abort of a transfer? What happens if the transfer is not in DONE state yet? If this call blocks, we shouldn't, instead return error. In destructor if we get to blocking state it's not end of the word, but in releaseReqH we should either abort, or return error in a non-blocking manner.

e-ago replied:
You can't abort a CUDA kernel. The kernel_wait polls on a flag updated by the CPU to exit.
This method is not really doing anything to the xfer request as it's all pre-allocated, anyway for correctness I've added a check on the input xfer request to check if it's in use or not.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+	}
+
+	if (msg.size() > DOCA_MAX_NOTIF_MESSAGE_SIZE - msg_tag_start.size() - msg_tag_end.size() - 16) {
+		std::cout 	<< "Can't send notif as message size " << msg.size()
```
Let's not have any cout s, use the debug print methods.

e-ago replied:
Replaced all std::cout and DOCA logs with NIXL logs

mkhazraee commented on src/utils/stream/metadata_stream.cpp:
Just a reminder, as we talked let's remove the changes from stream, making the PR smaller.

e-ago commented on test/nixl/nixl_doca_stream_test.cu:
Done

e-ago commented:
Changes in commits https://github.com/ai-dynamo/nixl/pull/404/commits/331c7c66eb048ea941a0fc01a4657ecaa969c9e1 and https://github.com/ai-dynamo/nixl/pull/404/commits/07b021ebbad442c396fdd7fa8ff4e1525448d39e :
- DOCA backend:
    - code cleanup, more readable now
    - All logs are now NIXL logs
- DOCA application
    - moved from test/nixl into test/unit/plugins
    - removed metadata streamer as OOB, now it uses new fetchRemote method
    - code quality improved
    - README

mkhazraee commented on test/README.md:
```diff
@@ -1,6 +1,6 @@
 # Test Descriptions
 
-Here are all the explained tests in this directory. There are more specific unit tests in src/utils.
+Here are all the explained tests in this directory. There are more specific unit tests in unit/plugins.
```
unit/plugins and unit/utils.

mkhazraee commented on test/unit/plugins/doca/README.md:
```diff
@@ -0,0 +1,53 @@
+
+It's possible to execute in two modes:
+- Stream attached: when creating a transfer request, the application can specify a CUDA stream where the transfer request (executed by a CUDA kernel implemented within the plugin) must be posted. Useful in situations where data is prepared in a GPU CUDA kernel and right after, asynchronously, it can be sent by the GPU without requiring any CPU synchronization.
+- Stream pool: at transfer request creation time, no CUDA stream is provided by the app. The DOCA plugin can still be used as it uses an internal pool of CUDA streams used in a round-robin fashion.
```
app -> application

mkhazraee commented on test/unit/plugins/doca/README.md:
```diff
@@ -0,0 +1,53 @@
+- Stream attached: when creating a transfer request, the application can specify a CUDA stream where the transfer request (executed by a CUDA kernel implemented within the plugin) must be posted. Useful in situations where data is prepared in a GPU CUDA kernel and right after, asynchronously, it can be sent by the GPU without requiring any CPU synchronization.
+- Stream pool: at transfer request creation time, no CUDA stream is provided by the app. The DOCA plugin can still be used as it uses an internal pool of CUDA streams used in a round-robin fashion.
+
+Please note in this unit test, data used for transfers is allocated in GPU memory. In real-world applications this is not mandatory: the DOCA plugin can send/receive data from the CUDA kernel if it resides in CPU memory.
```
Can we add such a test case too? If not that time consuming. And let's put a todo in the code and here mention that test will be added.

mkhazraee commented on test/unit/plugins/doca/README.md:
```diff
@@ -0,0 +1,53 @@
+
+## System requirements
+
+Please look at the [DOCA GPUNetIO](https://docs.nvidia.com/doca/sdk/doca+gpunetio/index.html) programming guide to properly configure youy system for the correct execution of this application.
```
youy -> your.
Does this mean running meson alone and the pacakges mentioned in readme is not sufficient? If we can detail more here it would be better, instead of sending them to another link, even though internal.

mkhazraee commented on test/unit/plugins/doca/README.md:
```diff
@@ -0,0 +1,53 @@
+
+## Output
+
+To ensure everything works correctly as expected, tagert should print out 32 lines like the following ones for the first transfer:
```
Maybe you can mention the flow, how they wait for notifications, and that things can be async and this is not mandatory, just the way the test is setup.

mkhazraee commented on src/plugins/doca/doca_plugin.cpp:
```diff
@@ -0,0 +1,83 @@
+
+// Static plugin structure
+static nixlBackendPlugin plugin = {NIXL_PLUGIN_API_VERSION, create_doca_engine,
+                                   destroy_doca_engine,     get_plugin_name,
```
Soon we will apply a code formatter, I think it would make these single spaced and not aligned. I don't recall seeing this anywhere else in the code, so let's be consistent.

e-ago replied:
fixed with clang-format using the config file you linked me

mkhazraee commented on src/plugins/doca/doca_backend_aux.h:
```diff
@@ -0,0 +1,210 @@
+#endif
+
+struct docaXferReqGpu {
+  uint32_t id;
```
Let's be consistent in the spacing, soon formatter will be applied, but before then let's use 4 spacing for a tab. I see you used 2 or 4 or 8.

mkhazraee replied:
I realized some places it was tabs, let's use spaces to avoid this.

mkhazraee commented on src/plugins/doca/doca_utils.cpp:
```diff
@@ -0,0 +1,282 @@
+	}
+}
+
+int oob_connection_client_setup(const char *server_ip, int *oob_sock_fd)
```
Note for @yosefe , when we move the MD manager separate and figure how we manage the stream util, maybe we should reuse that here, or keep this within the plugin to be standalone, just something to keep track of.

mkhazraee commented on src/plugins/doca/doca_backend_aux.h:
```diff
@@ -0,0 +1,210 @@
+
+#define DOCA_MAX_COMPLETION_INFLIGHT 64
+#define DOCA_DEVINFO_IBDEV_NAME_SIZE 64
+#define RDMA_RECV_QUEUE_SIZE 2048
```
This is for notifications, right? Let's do 4K if that doesn't cause cache issues.

mkhazraee replied:
Oh just saw the notif one, maybe we can add few comments for meaning of some of them. Or DOCA_XFER_REQ_SIZE. That's number of descs within a request?

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,193 @@
+
+  nixl_status_t checkXfer(nixlBackendReqH *handle) const override;
+  nixl_status_t releaseReqH(nixlBackendReqH *handle) const override;
+  int progress();
```
If you don't supprot progress, let's not provide it, same as loadLocal.

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,193 @@
+  void removeConnection(uint32_t connection_idx);
+  uint32_t getGpuCudaId();
+
+//   nixl_status_t sendLocalAgent(int oob_sock_client);
```
If not used anymore let's remove the comment. If these methods are internal to this backend, I guess making them private would be better. Same as the few methods at the beginning.

e-ago replied:
done

mkhazraee commented on src/plugins/doca/doca_backend.h:
```diff
@@ -0,0 +1,193 @@
+  nixl_status_t recvLocalAgent(int oob_sock_client, std::string &remote_agent);
+};
+
+doca_error_t doca_kernel_write(cudaStream_t stream,
```
If these 3 are helper functions, why not to only keep them in cpp file?

mkhazraee commented on src/plugins/doca/doca_utils.cpp:
```diff
@@ -0,0 +1,282 @@
+#include "doca_backend.h"
+#include "serdes/serdes.h"
+#include <arpa/inet.h>
+#include <cassert>
```
asserts can break the flow and we don't want them in the plugin. I see it's not used here, so please remove. I think it's fine for unit tests, anyways we want to fail.

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+	std::cout << "Starting Agent for "<< role << " with stream mode " << stream_mode << "\n";
+	/** Agent and backend creation parameters */
+
+	nixlAgentConfig cfg(true, true, 0, nixl_thread_sync_t::NIXL_THREAD_SYNC_STRICT);
```
Let's use peer port for both. 0 meant just use default port set by nixl, and that doesn't need to be tested here. Most probably we will actually change this to be consistent with linux, as 0 has a different meaning (OS assigned), so let's not have a dependency on it here.

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+	nixlSerDes *remote_serdes = new nixlSerDes();
+
+	/** Descriptors and Transfer Request */
+	nixl_reg_dlist_t  dram_for_doca(VRAM_SEG);
```
Since it's GPU it should be VRAM as we said, let's change the name vram_for_doca, or test_mem or so not to be confusing.

e-ago replied:
done

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+
+	/** NIXL declarations */
+	nixl_b_params_t params;
+	nixlBlobDesc    buf[TRANSFER_NUM_BUFFER];
```
I think we don't need an array, we just create one and use it right away in the code.

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+	}
+	/** Register memory in both initiator and target */
+	agent.registerMem(dram_for_doca, &extra_params);
+	agent.getLocalMD(metadata);
```
Since we're fetching, we don't need this getLocalMD or to put it in the notification message. It's already taken care of. Might be good to just check the size of output before you remote it for sanity checking.

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+
+		while (agent.genNotif(initiator, message, &extra_params) != NIXL_SUCCESS);
+
+		std::cout << " Sending serdes to " << initiator << " to enable notifications" << std::endl;
```
I think this should be before the while, and maybe make it more clear, sending buffer information to the initiator. (Nore more sending metadata due to fetch, and even then enable notifications might be confusing).

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+		md_extra_params.port = peer_port;
+
+		ret = agent.fetchRemoteMD(target, &md_extra_params);
+		std::cout << "fetchRemoteMD returned " << ret << " to " << target << std::endl;
```
fetchRemoteMD might return error, until the data is received, we should do a fetch, and then checkRemoteMD. So better to first do send, then fetch, then wait for checkRemoteMD.

mkhazraee replied:
Since you did the send and then waited for notif, this issue was hidden. Another option is to put the checkRemoteMD after the getNotif, not to block while waiting for the notification, not much a difference.

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+			buf[i].addr  = (uintptr_t)(data_address_ptr + (i * buf_size));
+			buf[i].len   = buf_size;
+			buf[i].devId = 0;
+			dram_target_doca.addDesc(buf[i]);
```
let's rename not to have dram.

mkhazraee commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+	std::cout << "Cleanup.. \n";
+	
+	//do we need extra_params?
+	//Check if deregistration also works implicitely when closing the app without this function
```
reminder.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	}
+	NIXL_INFO << std::endl;
+
+	if (custom_params->count("gpu_devices") == 0)
```
Why can't we just get the cudacontext during the first register like UCX? This way user can just get a memory from torch and pass it down, versus finding the device after getting the memory, and then initializing the backend. I'm borderline if we should make it for current release, as it breaks the python example flows and makes adoption hard.

e-ago replied:
I can get device from CUDA context but this may not be ok with other use-cases. Maybe we should enable some special value for gpu_devices to enable this "retrieve cuda id from context" mode

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	memset(notif_send_cpu, 0, sizeof(struct docaNotifSend) * 1);
+
+	cuCtxGetCurrent(&main_cuda_ctx);
+	// Warmup
```
How long does a call to this constructor take? If it's long we should really think about using a progress thread in the next release. Warming up doesn't sound something we should do in constructor.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	req->~nixlDocaBckndReq();
+}
+
+nixl_status_t nixlDocaEngine::progressThreadStart() {
```
I'm confused, why supportsProgTh passes false then?

mkhazraee replied:
I see it's just the connection setup, but def it can have other use cases, for this release let's say progress thread is not complete, todo to add more features and then set it to true. Something like
// For now just connection setup, not used for xfers to be a complete progThread, so supportsProgTh is false

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+
+	/* Set permissions to DOCA RDMA */
+	result = doca_rdma_set_permissions(rdma_qp->rdma_data,
+										DOCA_ACCESS_FLAG_LOCAL_READ_WRITE |
```
tab misalignment, let's use spaces all over.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	return NIXL_SUCCESS;
+}
+
+nixl_status_t nixlDocaEngine::recvLocalAgent(int oob_sock_client,
```
What is the role of this method? to get QP info for local and remote?

e-ago replied:
nice catch fixed names and arguments, thanks!

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	return NIXL_SUCCESS;
+}
+
+nixl_status_t nixlDocaEngine::loadLocalMD(nixlBackendMD *input,
```
I thought I've put a comment, let's remove this method, as the flag indicates it is not supported, the virtual method doesn't need to be provided in this scenario.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	if (((volatile docaXferCompletion *)completion_list_cpu)[tmp].completed > 0)
+		return NIXL_SUCCESS;
+	else
+		return NIXL_IN_PROG;
```
If it cannot be terminated, maybe we should return error. Good for now, but a decision we should make @yosefe

mkhazraee replied:
Ran it by Yossi, closing it for now.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+		return NIXL_IN_PROG;
+}
+
+int nixlDocaEngine::progress() { return NIXL_SUCCESS; }
```
Not to be implemented if the flag is set to false.

e-ago replied:
removed

e-ago commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1454 @@
+	((volatile uint8_t *)wait_exit_cpu)[0] = 1;
+	cudaStreamSynchronize(wait_stream);
+	cudaStreamDestroy(wait_stream);
+	NIXL_ERROR << "free wait_exit_gpu";
```
removed

e-ago commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+	if (role == target) {
+		bool found = false;
+		//Not used
+		nixl_xfer_dlist_t descs(DRAM_SEG);
```
done

e-ago commented on test/unit/plugins/doca/nixl_doca_stream_test.cu:
```diff
@@ -0,0 +1,603 @@
+		dram_for_doca.addDesc(buf[i]);
+	}
+	/** Register memory in both initiator and target */
+	agent.registerMem(dram_for_doca, &extra_params);
```
done

mkhazraee commented:
/ok to test 8c83d30

mkhazraee commented on src/utils/stream/metadata_stream.cpp:
```diff
@@ -199,7 +199,6 @@ bool nixlMDStreamClient::connectListener() {
    return setupClient();
 }
 
```
Let's remove this space removal, so the file doesn't show up in the PR.

mkhazraee commented on src/plugins/doca/doca_backend.cpp:
```diff
@@ -0,0 +1,1868 @@
+
+int nixlDocaEngine::progress() { return NIXL_SUCCESS; }
+
+nixl_status_t nixlDocaEngine::getNotifs(notif_list_t &notif_list) {
```
@e-ago not for this release, but just should we consider this for the next release?

mkhazraee commented:
/ok to test b09afee

rakhmets commented on src/plugins/gpunetio/gpunetio_backend_aux.h:
```diff
@@ -0,0 +1,230 @@
+#define DOCA_MAX_NOTIF_INFLIGHT RDMA_RECV_QUEUE_SIZE
+#define DOCA_MAX_NOTIF_MESSAGE_SIZE 8192
+#define DOCA_NOTIF_NULL 0xFFFFFFFF
+#define DOCA_MSG_TAG 0xFF
```
Please move defines that used only in a .cpp file to this file. And consider using constexpr instead of define.

rakhmets commented on src/plugins/gpunetio/gpunetio_utils.cpp:
```diff
@@ -0,0 +1,256 @@
+    size_t i;
+
+    /* Set default return value */
+    *retval = NULL;
```
Please replace NULL with nullptr.

rakhmets commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+        goto exit_error;
+    }
+
+    // /* Set gid_index to DOCA RDMA if it's provided */
+    // if (cfg->is_gid_index_set) {
+    // 	/* Set gid_index to DOCA RDMA */
+    // 	result = doca_rdma_set_gid_index(rdma, cfg->gid_index);
+    // 	if (result != DOCA_SUCCESS) {
+    // 		NIXL_ERROR << "Failed to set gid_index to DOCA RDMA " << // doca_error_get_descr(result);
+    // goto exit_error;
+    // 	}
+    // }
```
Please remove unused code.

rakhmets commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+nixl_status_t
+nixlDocaEngine::recvRemoteAgentName (int oob_sock_client, std::string &remote_agent) {
+    size_t msg_size;
+    char *remote_agent_;
```
This is not needed. Use resize method of the string, and pass remote_agent.data() to recv.

yosefe commented:
see AWS failure:
```
Run-time dependency etcd-cpp-api found: NO (tried pkgconfig and cmake)
Message: ETCD CPP API library not found, will disable etcd support
Program python3 found: YES (/usr/bin/python3)
Library aio found: YES
Run-time dependency liburing found: NO (tried pkgconfig and cmake)
Checking for function "aio_cancel" : YES 
Message: Correct libaio found, Building POSIX plugin
Message: liburing not found, building with AIO support only
Library transfer_engine found: NO
Library hf3fs_api_shared found: NO
Run-time dependency doca-gpunetio found: NO (tried pkgconfig and cmake)

src/plugins/gpunetio/meson.build:16:23: ERROR: Dependency "doca-gpunetio" not found, tried pkgconfig and cmake
```

e-ago commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+
+    memset (notif_send_cpu, 0, sizeof (struct docaNotifSend) * 1);
+
+    cuCtxGetCurrent (&main_cuda_ctx);
```
done

e-ago commented on src/plugins/gpunetio/gpunetio_backend.h:
```diff
@@ -0,0 +1,238 @@
+    recvRemoteAgentName (int oob_sock_client, std::string &remote_agent);
+};
+
+doca_error_t
+doca_kernel_write (cudaStream_t stream,
+                   struct doca_gpu_dev_rdma *rdma_gpu,
+                   struct docaXferReqGpu *xferReqRing,
+                   uint32_t pos);
+doca_error_t
+doca_kernel_read (cudaStream_t stream,
+                  struct doca_gpu_dev_rdma *rdma_gpu,
+                  struct docaXferReqGpu *xferReqRing,
+                  uint32_t pos);
+doca_error_t
+doca_kernel_progress (cudaStream_t stream,
+                      struct docaXferCompletion *completion_list,
+                      struct docaNotifRecv *notif_fill,
+                      struct docaNotifRecv *notif_progress,
+                      struct docaNotifSend *notif_send_gpu,
+                      uint32_t *exit_flag);
```
moved into `gpunetio_backend_aux.h`

e-ago commented on src/plugins/gpunetio/gpunetio_backend_aux.h:
```diff
@@ -0,0 +1,230 @@
+#define DOCA_ENG_MAX_CONN 20
+#define DOCA_RDMA_CM_LOCAL_PORT_CLIENT 6543
+#define DOCA_RDMA_CM_LOCAL_PORT_SERVER 6544
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
```
used later for another macro

rakhmets replied:
DOCA_RDMA_SERVER_ADDR_LEN seems to be unused. Please remove.

aranadive replied:
Thanks for pointing this out Raul. Lets have a separate cleanup PR.

e-ago commented on src/plugins/gpunetio/gpunetio_backend_aux.h:
```diff
@@ -0,0 +1,230 @@
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+#define DOCA_RDMA_SERVER_ADDR_LEN \
+    (MAX (MAX (DOCA_DEVINFO_IPV4_ADDR_SIZE, DOCA_DEVINFO_IPV6_ADDR_SIZE), DOCA_GID_BYTE_LENGTH))
+#define DOCA_RDMA_SERVER_CONN_DELAY 500 // 500us
```
done

e-ago commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+
+    char pciBusId[DOCA_DEVINFO_IBDEV_NAME_SIZE];
+    for (auto &item : gdevs) {
+        cudaDeviceGetPCIBusId (pciBusId, DOCA_DEVINFO_IBDEV_NAME_SIZE, item.first);
```
done

e-ago commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+        treq->stream = (cudaStream_t) * ((uintptr_t *)opt_args->customParam.data());
+    }
+
+#if 0
```
done

e-ago commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+                                 DOCA_GPU_MEM_TYPE_GPU,
+                                 (void **)&last_rsvd_flags,
+                                 nullptr);
+    if (result != DOCA_SUCCESS || last_rsvd_flags == NULL || last_rsvd_flags == NULL) {
```
done

e-ago commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+                                  "Failed to memset GPU memory");
+
+    result = doca_gpu_mem_alloc (gdevs[0].second,
+                                 sizeof (uint64_t) * 1,
```
applied locally

e-ago commented on src/plugins/gpunetio/gpunetio_kernels.cu:
```diff
@@ -0,0 +1,453 @@
+
+    /* Check no previous CUDA errors */
+    result = cudaGetLastError();
+    if (result != cudaSuccess) {
+        fprintf (stderr,
+                 "[%s:%d] cuda failed with %s",
+                 __FILE__,
+                 __LINE__,
+                 cudaGetErrorString (result));
+        return DOCA_ERROR_BAD_STATE;
+    }
```
it's just local in this file

e-ago commented on src/plugins/gpunetio/gpunetio_backend.cpp:
```diff
@@ -0,0 +1,1575 @@
+        NIXL_ERROR << "Function doca_gpu_mem_alloc return " << doca_error_get_descr (result);
+    }
+
+    nixlDocaEngineCheckCudaError (cudaMemset (last_rsvd_flags, 0, sizeof (uint64_t) * 1),
```
done locally

e-ago commented:
> see AWS failure:
> 
> ```
> Run-time dependency etcd-cpp-api found: NO (tried pkgconfig and cmake)
> Message: ETCD CPP API library not found, will disable etcd support
> Program python3 found: YES (/usr/bin/python3)
> Library aio found: YES
> Run-time dependency liburing found: NO (tried pkgconfig and cmake)
> Checking for function "aio_cancel" : YES 
> Message: Correct libaio found, Building POSIX plugin
> Message: liburing not found, building with AIO support only
> Library transfer_engine found: NO
> Library hf3fs_api_shared found: NO
> Run-time dependency doca-gpunetio found: NO (tried pkgconfig and cmake)
> 
> src/plugins/gpunetio/meson.build:16:23: ERROR: Dependency "doca-gpunetio" not found, tried pkgconfig and cmake
> ```

How should I fix it?

aranadive commented on src/plugins/meson.build:
```diff
@@ -37,3 +37,8 @@ hf3fs_lib_found = cc.find_library(hf3fs_lib_file, dirs: [hf3fs_lib_path], requir
     subdir('hf3fs')
 endif
+
+disable_gpunetio_backend = get_option('disable_gpunetio_backend')
```
Also check for the doca-gpunetio dependency here. If not found, no need to process the gpunetio dir. See the mooncake check above. This should help with the AWS error. If doca-gpunetio is part of CUDA, please check it here as well.

e-ago commented on src/plugins/gpunetio/gpunetio_backend_aux.h:
```diff
@@ -0,0 +1,230 @@
+                     struct doca_buf_arr **barr,
+                     struct doca_gpu_buf_arr **barr_gpu);
+doca_error_t
+destroy_doca_buf_arr (struct doca_buf_arr *barr);
```
done

aranadive commented on meson.build:
```diff
@@ -54,6 +54,29 @@ else
+endif
+
+# DOCA
+doca_gpunetio_dep = dependency('doca-gpunetio', required : true)
```
this should be required : false.

aranadive commented:
/ok to test 57ee683

aranadive commented:
/ok to test 57ee683

aranadive commented on src/core/nixl_agent.cpp:
```diff
@@ -742,6 +742,8 @@ nixlAgent::createXferReq(const nixl_xfer_op_t &operation,
         return NIXL_ERR_BACKEND;
     }
 
+    opt_args.customParam = extra_params->customParam;
```
Fixed

aranadive commented on src/plugins/gpunetio/gpunetio_backend.h:
```diff
@@ -0,0 +1,238 @@
+    struct docaNotifRecv *notif_progress_cpu;
+
+    struct docaNotifSend *notif_send_gpu;
+    struct docaNotifSend *notif_send_cpu;
```
Resolve in new PR.

aranadive commented:
/ok to test 10707cb

aranadive commented:
/build

aranadive commented:
/ok to test 46e16c6

aranadive commented:
/build

aranadive commented:
/ok to test b1da134

aranadive commented:
/build

aranadive commented:
/build

rakhmets commented on src/plugins/gpunetio/gpunetio_backend.h:
```diff
@@ -0,0 +1,238 @@
+
+class nixlDocaEngine : public nixlBackendEngine {
+private:
+    struct doca_log_backend *sdk_log;
```
Because it's redundant in c++.

e-ago commented on meson.build:
```diff
@@ -54,6 +54,29 @@ else
+    nvcc = meson.get_compiler('cuda')
+
+    nvcc_flags = []
+    nvcc_flags += ['-gencode', 'arch=compute_80,code=sm_80']
+    nvcc_flags += ['-gencode', 'arch=compute_90,code=sm_90']
```
added a comment

yosefe commented on contrib/Dockerfile:
```diff
@@ -109,6 +109,20 @@ RUN cd /usr/local/src && \
+    tee /etc/apt/sources.list.d/mellanox.list && \
+    DEBIAN_FRONTEND=noninteractive apt update -y && \
+    apt install -y --no-install-recommends \
+        mlnx-dpdk mlnx-dpdk-dev \
```
So if there is a dependency why need to install it explicitly, can't apt install resolve it?

aranadive replied:
I also noticed that this list installs ucx again. Can we move this doca install block to before UCX build and install? Move code from lines 112-125 to line 87 above? My mistake I didnt notice before.

yosefe replied:
install ucx from source after this step to overwrite ucvx from doca. add big comment in this file to explain.

yosefe commented on src/api/cpp/backend/backend_aux.h:
```diff
@@ -32,6 +32,7 @@ struct nixlBackendOptionalArgs {
     // During postXfer, user might ask for a notification if supported
     nixl_blob_t notifMsg;
     bool        hasNotif = false;
+    nixl_blob_t customParam;
```
@mkhazraee IMO "nixl_blob_t gpu_stream" is generic if another plugin needs gpu stream, customParam opens the door for ugliness like customParam2, customParam3, newCustomParam etc

aranadive replied:
It can be called customParams to indicate multiple params can be passed. gpu_stream seems specific here. We can pass something like num of queues to use for posix io_uring here, and gpu_stream would not make sense.

yosefe commented on src/core/nixl_agent.cpp:
```diff
@@ -732,9 +732,14 @@ nixlAgent::createXferReq(const nixl_xfer_op_t &operation,
+            opt_args.hasNotif = true;
+        }
+
+        if (extra_params->customParam.length() > 0)
```
can we remove?

aranadive commented on contrib/build-container.sh:
```diff
@@ -150,14 +150,15 @@ show_build_options() {
 show_help() {
     echo "usage: build-container.sh"
-    echo "  [--base base image]"
+    echo "  [--base-image base image]"
```
Where is it used? This is just update the command line correctly. CI seemed to be okay with the change.

e-ago commented:
@yosefe  is the commit https://github.com/ai-dynamo/nixl/pull/404/commits/4086d89b65d6ad3a9b3f12acbf71cfafcedd72c2 ok?

yosefe commented:
/build
